# Feature engineering for each participant in our experiment

Since we have problems regarding with complementary information between equipments, we will drop GSR measurements and focus on facial equipment only

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

%matplotlib inline
pd.set_option('display.max_columns', None)

In [ ]:
# Read csv file

df = pd.read_csv(r"C:\Users\berny\Documents\tiktok\5.1 IAD TikTok-1\Sensor Data\001_Martin Garcia Torres.csv", skiprows=31)
df.head()


C:\Users\berny\AppData\Local\Temp\ipykernel_28216\1910370799.py:3: DtypeWarning: Columns (3,4,6) have mixed types. Specify dtype option on import or set low_memory=False.
  original_df = pd.read_csv(r"C:\Users\berny\Documents\tiktok\5.1 IAD TikTok-1\Sensor Data\001_Martin Garcia Torres.csv", skiprows=31)


Row  Timestamp  EventSource  SlideEvent   StimType  Duration  \
0    1    32.3639          1.0  StartSlide  TestImage  900000.0   
1    2    35.6886          NaN         NaN        NaN       NaN   
2    3    43.5011          NaN         NaN        NaN       NaN   
3    4    51.3136          NaN         NaN        NaN       NaN   
4    5    59.1261          NaN         NaN        NaN       NaN   

  CollectionPhase   SourceStimuliName  EventSource.1  SampleNumber  Anger  \
0  StimuliDisplay  Screen recording 1            NaN           NaN    NaN   
1             NaN  Screen recording 1            NaN           NaN    NaN   
2             NaN  Screen recording 1            NaN           NaN    NaN   
3             NaN  Screen recording 1            NaN           NaN    NaN   
4             NaN  Screen recording 1            NaN           NaN    NaN   

   Contempt  Disgust  Fear  Joy  Sadness  Surprise  Engagement  \
0       NaN      NaN   NaN  NaN      NaN       NaN         NaN   
1       NaN      NaN   NaN  NaN      NaN       NaN         NaN   
2       NaN      NaN   NaN  NaN      NaN       NaN         NaN   
3       NaN      NaN   NaN  NaN      NaN       NaN         NaN   
4       NaN      NaN   NaN  NaN      NaN       NaN         NaN   

   Adaptive Engagement  Valence  Adaptive Valence  Sentimentality  Confusion  \
0                  NaN      NaN               NaN             NaN        NaN   
1                  NaN      NaN               NaN             NaN        NaN   
2                  NaN      NaN               NaN             NaN        NaN   
3                  NaN      NaN               NaN             NaN        NaN   
4                  NaN      NaN               NaN             NaN        NaN   

   Neutral  Attention  Brow Furrow  Brow Raise  Cheek Raise  Chin Raise  \
0      NaN        NaN          NaN         NaN          NaN         NaN   
1      NaN        NaN          NaN         NaN          NaN         NaN   
2      NaN        NaN          NaN         NaN          NaN         NaN   
3      NaN        NaN          NaN         NaN          NaN         NaN   
4      NaN        NaN          NaN         NaN          NaN         NaN   

   Dimpler  Eye Closure  Eye Widen  Inner Brow Raise  Jaw Drop  \
0      NaN          NaN        NaN               NaN       NaN   
1      NaN          NaN        NaN               NaN       NaN   
2      NaN          NaN        NaN               NaN       NaN   
3      NaN          NaN        NaN               NaN       NaN   
4      NaN          NaN        NaN               NaN       NaN   

   Lip Corner Depressor  Lip Press  Lip Pucker  Lip Stretch  Lip Suck  \
0                   NaN        NaN         NaN          NaN       NaN   
1                   NaN        NaN         NaN          NaN       NaN   
2                   NaN        NaN         NaN          NaN       NaN   
3                   NaN        NaN         NaN          NaN       NaN   
4                   NaN        NaN         NaN          NaN       NaN   

   Lid Tighten  Mouth Open  Nose Wrinkle  Smile  Smirk  Upper Lip Raise  \
0          NaN         NaN           NaN    NaN    NaN              NaN   
1          NaN         NaN           NaN    NaN    NaN              NaN   
2          NaN         NaN           NaN    NaN    NaN              NaN   
3          NaN         NaN           NaN    NaN    NaN              NaN   
4          NaN         NaN           NaN    NaN    NaN              NaN   

   Blink  BlinkRate  Speaking  Pitch  Yaw  Roll  Interocular Distance  \
0    NaN        NaN       NaN    NaN  NaN   NaN                   NaN   
1    NaN        NaN       NaN    NaN  NaN   NaN                   NaN   
2    NaN        NaN       NaN    NaN  NaN   NaN                   NaN   
3    NaN        NaN       NaN    NaN  NaN   NaN                   NaN   
4    NaN        NaN       NaN    NaN  NaN   NaN                   NaN   

   EventSource.2  SampleNumber.1  width  height  feature id_1  feature-x_1  \
0   

In [ ]:
# Drop gsr features
gsr_features = [feature for feature in df.columns if df.columns.get_loc(feature) > 130]

df.drop(columns=gsr_features, axis=1, inplace=True)

In [ ]:
ignore_cols = [feature for feature in df.columns if feature == 'Rows' or feature == 'Timestamp' or feature == '']

In [3]:
features_with_nan = [feature for feature in original_df.columns if original_df[feature].isnull().sum() > 0]

for feature in features_with_nan:
    print(feature, np.round(original_df[feature].isnull().mean(), 4) * 100, '% missing values')

EventSource 100.0 % missing values
SlideEvent 100.0 % missing values
StimType 100.0 % missing values
Duration 100.0 % missing values
CollectionPhase 100.0 % missing values
EventSource.1 81.14 % missing values
SampleNumber 81.14 % missing values
Anger 81.14 % missing values
Contempt 81.14 % missing values
Disgust 81.14 % missing values
Fear 81.14 % missing values
Joy 81.14 % missing values
Sadness 81.14 % missing values
Surprise 81.14 % missing values
Engagement 81.14 % missing values
Adaptive Engagement 81.14 % missing values
Valence 81.14 % missing values
Adaptive Valence 81.14 % missing values
Sentimentality 81.14 % missing values
Confusion 81.14 % missing values
Neutral 81.14 % missing values
Attention 81.14 % missing values
Brow Furrow 81.14 % missing values
Brow Raise 81.14 % missing values
Cheek Raise 81.14 % missing values
Chin Raise 81.14 % missing values
Dimpler 81.14 % missing values
Eye Closure 81.14 % missing values
Eye Widen 81.14 % missing values
Inner Brow Raise 81.14 % 

We can see a pattern of 81.14% of missing values, this insight is useful because probably there is a constant timestamp where values are updated, we need to find that timestamp

Also there are some features with 100% of missing values, we should get rid of those

Finally another pattern of 19.72% and 19.48% of missing values.

For the purpose of this analysis, we are going to drop all columns that carry no information.

It is important to note that for a linear regression model, the dropping of rows carry no consequences but if we were to apply any sequential model like ARIMA or an LSTM, then removing rows will break the regular cadence.

We have not decided on a model, so for now we will be performing imputation techniques in order to preserve the timestamp

If I were to drop missing information rows and then decided to go for a sequential model, then it is recommended to full re-index the dropped timestamps and apply some imputation technique on the missing rows. s

After further analysis:

- 81% of missing info belongs to the facial equipment
- 19% of missing info belongs to the GSR equipment

GSR keeps the majority of information constant, probably with some interpolation technique.

Ideas?
1. Interpolate data from the facial equipment and interpolate missing data from GSR equipment
2. Drop missing information from the facial equipment, even though we may delete some rows regarding GSR information

The objective is to have both measurement equipment unified in a single csv in order to perform Exploratory Data Analysis.

First we will perform an imputation technique with all values regarding GSR equipment, we will impute with the median of all values, since the majority of the distributions are not close to a normal distribution

Then we will drop all column that give no information about anything, or are exactly the same as others, thus we avoid multicollinearity

Finally we will impute the median values to all Facial Equipment columns, since they are not close to a normal distribution

In [4]:
columns_to_impute = list(original_df.iloc[:, 120:].columns)

for cols in columns_to_impute:
    median = original_df[cols].median()
    original_df[cols].fillna(median, inplace=True)

C:\Users\berny\AppData\Local\Temp\ipykernel_28216\320876554.py:5: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  original_df[cols].fillna(median, inplace=True)


In [5]:
# First let us drop those columns with 100% of missing values
# And also drop the ones that just repeat themselves.

original_df.drop(['EventSource', 'SlideEvent', 'StimType', 'Duration', 'CollectionPhase', 'EventSource.3', 'CPU Sys', 'Memory Sys', 'CPU Proc', 'Memory Proc', 
                  'EventSource.1', 'SampleNumber', 'Valence', 'EventSource.2', 'SampleNumber.1', 'width', 'height', 'feature id_1', 'feature id_2', 'feature id_3',
                  'feature id_4', 'feature id_5', 'feature id_6', 'EventSource.4', 'Anger.1', 'Contempt.1', 'Disgust.1', 'Fear.1', 'Joy.1', 'Sadness.1', 'Surprise.1', 
                  'Engagement.1', 'Adaptive Engagement.1', 'Negative Adaptive Valence', 'Neutral Adaptive Valence', 'Sentimentality.1', 'Nose Wrinkle.1', 'EventSource.5', 
                  'EventSource.6', 'SampleNumber.2', 'Timestamp RAW', 'Timestamp CAL', 'System Timestamp CAL', 'VSenseBatt RAW', 'Packet reception rate RAW', 'Negative Valence'], axis=1, inplace=True)

In [38]:
from sklearn.impute import KNNImputer
from sklearn.preprocessing import StandardScaler

# 1. Load your data
df = original_df

# 2. Identify columns
cont_cols = list(df.columns[3:56])   # continuous features
disc_cols = list(df.columns[56:93])  # discrete (0/1) features

# 3. Choose neighbour features
neigh = []
if 'GSR Raw' in df.columns:
    neigh.append('GSR Raw')

# 4. Build the matrix to impute
impute_cols = cont_cols + disc_cols + neigh
X = df[impute_cols].copy()

# 6. Scale everything
scaler = StandardScaler()
X_scaled = scaler.fit_transform(X)

# 7. KNN impute
imputer = KNNImputer(n_neighbors=5, weights='uniform')
X_filled = imputer.fit_transform(X_scaled)

# 8. Inverse‐scale
X_inv = scaler.inverse_transform(X_filled)
X_imputed = pd.DataFrame(X_inv, columns=impute_cols, index=df.index)

# 9. Put imputed values back into your original df
df_imputed = df.copy()

#   – continuous: use as is
df_imputed[cont_cols] = X_imputed[cont_cols]

#   – discrete: round to nearest integer (0 or 1) and clip
df_imputed[disc_cols] = (
    X_imputed[disc_cols]
      .round()
      .clip(lower=0, upper=1)
      .astype(int)
)

df_imputed

# Now df_imputed has your KNN‐filled features


KeyboardInterrupt: 

In [36]:
from sklearn.impute import KNNImputer
from sklearn.preprocessing import StandardScaler
import time

# 1. Load your data
df = original_df

# 2. Identify columns
cont_cols = list(df.columns[3:56])   # continuous features
disc_cols = list(df.columns[56:93])  # discrete (0/1) features

# 3. Choose neighbour features
#    Prefer 'Timestamp' if present, otherwise 'GSR Raw' (or both).
neigh = []
if 'Timestamp' in df.columns:
    neigh.append('Timestamp')
if 'GSR Raw' in df.columns:
    neigh.append('GSR Raw')

# 4. Build the matrix to impute
impute_cols = cont_cols + disc_cols + neigh
X = df[impute_cols].copy()

# 6. Scale everything
scaler = StandardScaler()
X_scaled = scaler.fit_transform(X)


# X_scaled is your full scaled matrix (n_rows × n_feats)
sample_size = 10_000
sample = X_scaled[:sample_size]

imputer = KNNImputer(n_neighbors=5, weights='uniform')

# Time the 10k-row sample
start = time.time()
imputer.fit_transform(sample)
t_sample = time.time() - start

# Extrapolate to full size
n_rows = X_scaled.shape[0]
t_est = t_sample * (n_rows / sample_size)

print(f"{t_sample:.2f}s for {sample_size} rows → estimated {t_est/60:.1f} min for full dataset")

# Now df_imputed has your KNN‐filled features


64.56s for 10000 rows → estimated 15.4 min for full dataset


In [ ]:
original_df.head()

Row  Timestamp   SourceStimuliName  Anger  Contempt  Disgust  Fear  Joy  \
0    1    32.3639  Screen recording 1    NaN       NaN      NaN   NaN  NaN   
1    2    35.6886  Screen recording 1    NaN       NaN      NaN   NaN  NaN   
2    3    43.5011  Screen recording 1    NaN       NaN      NaN   NaN  NaN   
3    4    51.3136  Screen recording 1    NaN       NaN      NaN   NaN  NaN   
4    5    59.1261  Screen recording 1    NaN       NaN      NaN   NaN  NaN   

   Sadness  Surprise  Engagement  Adaptive Engagement  Adaptive Valence  \
0      NaN       NaN         NaN                  NaN               NaN   
1      NaN       NaN         NaN                  NaN               NaN   
2      NaN       NaN         NaN                  NaN               NaN   
3      NaN       NaN         NaN                  NaN               NaN   
4      NaN       NaN         NaN                  NaN               NaN   

   Sentimentality  Confusion  Neutral  Attention  Brow Furrow  Brow Raise  \
0             NaN        NaN      NaN        NaN          NaN         NaN   
1             NaN        NaN      NaN        NaN          NaN         NaN   
2             NaN        NaN      NaN        NaN          NaN         NaN   
3             NaN        NaN      NaN        NaN          NaN         NaN   
4             NaN        NaN      NaN        NaN          NaN         NaN   

   Cheek Raise  Chin Raise  Dimpler  Eye Closure  Eye Widen  Inner Brow Raise  \
0          NaN         NaN      NaN          NaN        NaN               NaN   
1          NaN         NaN      NaN          NaN        NaN               NaN   
2          NaN         NaN      NaN          NaN        NaN               NaN   
3          NaN         NaN      NaN          NaN        NaN               NaN   
4          NaN         NaN      NaN          NaN        NaN               NaN   

   Jaw Drop  Lip Corner Depressor  Lip Press  Lip Pucker  Lip Stretch  \
0       NaN                   NaN        NaN         NaN          NaN   
1       NaN                   NaN        NaN         NaN          NaN   
2       NaN                   NaN        NaN         NaN          NaN   
3       NaN                   NaN        NaN         NaN          NaN   
4       NaN                   NaN        NaN         NaN          NaN   

   Lip Suck  Lid Tighten  Mouth Open  Nose Wrinkle  Smile  Smirk  \
0       NaN          NaN         NaN           NaN    NaN    NaN   
1       NaN          NaN         NaN           NaN    NaN    NaN   
2       NaN          NaN         NaN           NaN    NaN    NaN   
3       NaN          NaN         NaN           NaN    NaN    NaN   
4       NaN          NaN         NaN           NaN    NaN    NaN   

   Upper Lip Raise  Blink  BlinkRate  Speaking  Pitch  Yaw  Roll  \
0              NaN    NaN        NaN       NaN    NaN  NaN   NaN   
1              NaN    NaN        NaN       NaN    NaN  NaN   NaN   
2              NaN    NaN        NaN       NaN    NaN  NaN   NaN   
3              NaN    NaN        NaN       NaN    NaN  NaN   NaN   
4              NaN    NaN        NaN       NaN    NaN  NaN   NaN   

   Interocular Distance  feature-x_1  feature-y_1  feature-x_2  feature-y_2  \
0                   NaN          NaN          NaN          NaN          NaN   
1                   NaN          NaN          NaN          NaN          NaN   
2                   NaN          NaN          NaN          NaN          NaN   
3                   NaN          NaN          NaN          NaN          NaN   
4                   NaN          NaN          NaN          NaN          NaN   

   feature-x_3  feature-y_3  feature-x_4  feature-y_4  feature-x_5  \
0          NaN          NaN          NaN          NaN          NaN   
1          NaN          NaN          NaN          NaN          NaN   
2          NaN          NaN          NaN          NaN          NaN   
3          NaN          NaN          NaN          NaN          NaN   
4          NaN          NaN          NaN          N

In [ ]:
possible_dependent_features = [feature for feature in df.columns if feature.startswith('Enga')]
print(possible_dependent_features)

What is the difference between `Engagement` and `Engagement.1` ?

In [ ]:
dependent_features = df[['Engagement', 'Engagement.1']]
dependent_features

`Engagement` is a continuous value from 0 to 1 that quantifies how much the person is engaged, this leads to the possibility of a linear regression model

`Engagement.1` is a discrete value going either 0 or 1 that its dependant of `Engagement` I assume using a threshold.

- For example if the feature `Engagement` is greater than 0.5, then `Engagement.1` is 1, else is 0.

If we count `Engagement.1` as a dependant feature then we might use a logistic regression model. 

For now let us use `Engagement` as our dependent feature